lib/charms/ip_router_interface/v0/ip_router_interface.py

```diff
@@ -27,6 +136,312 @@
+    """
+
+    on = RouterProviderCharmEvents()
+
+    def __init__(self, charm: CharmBase, relationship_name: str = "ip-router"):
+        super().__init__(charm, relationship_name)
+        self.charm = charm
+        self.relationship_name = relationship_name
+        self.framework.observe(
+            charm.on[relationship_name].relation_changed, self._router_relation_changed
+        )
+        self.framework.observe(
+            charm.on[relationship_name].relation_departed, self._router_relation_departed
+        )
+
+    def _router_relation_changed(self, event: RelationChangedEvent):
+        """Resync the databags since there could have been a change in networks."""
+        if not self.charm.unit.is_leader():
+            return
+        self._sync_routing_tables()
+        self.on.routing_table_updated.emit()
+
+    def _router_relation_departed(self, event: RelationDepartedEvent):
+        """If an application has completely departed the relation, remove it
+        from the routing table.
+        """
+        if not self.charm.unit.is_leader():
+            return
+        self._sync_routing_tables()
+
+    def get_routing_table(self) -> RoutingTable:
+        """Build the routing table from all of the related databags. Relations
+        that don't have missing or invalid network requests will be ignored.
+        """
+        logger.debug("Routing Table generation starting.")
+        router_relations = self.model.relations[self.relationship_name]
+        final_routing_table: RoutingTable = {}
+        for relation in router_relations:
+            new_network_name = relation.data[relation.app].get("network-name", None)
+            new_network_request: List[Network] = json.loads(
+                relation.data[relation.app].get("networks", "{}")
+            )
+
+            logger.debug(
+                "Attempting to add (%s) from app:(%s) with relation-name:(%s)",
+                new_network_request,
+                relation.app.name,
+                new_network_name,
+            )
+
+            if (
+                not new_network_name
+                or not new_network_request
+                or _network_name_taken(new_network_name, router_relations)
+            ):
+                continue
+
+            final_routing_table[new_network_name] = []
+
+            for network in new_network_request:
+                try:
+                    _validate_network(network, final_routing_table)
+                except (ValueError, KeyError) as e:
+                    logger.error(
+                        "Exception (%s) occurred with network %s. Skipping this entry.",
+                        e.args[0],
+                        network,
+                    )
+                else:
+                    final_routing_table[new_network_name].append(network)
+
+        logger.debug("Generated rt: %s", final_routing_table)
+        return final_routing_table
+
+    def get_flattened_routing_table(self) -> List[Network]:
```

Comment:
Since routing table is an object, couldn't it have a "flatten" method?

```python
table = RoutingTable()
print(table.flatten())
```

Reply:
RoutingTable is currently a type alias, for Dict[str, ...], and as far as I know it's not possible to add methods to built-in primitives in Python. I can make RoutingTable into an object if you like.

Reply:
Maybe we could use Pydantic here as the model instead of TypeAliases. This would give us complete objects where we can add functionality, and we could use Pydantic for the validation. A good example is there: https://github.com/canonical/traefik-k8s-operator/blob/main/lib/charms/traefik_k8s/v2/ingress.py

Reply:
I love Pydantic :D It's something I've thought of from the beginning, but I wasn't sure how to make sure to have it be included everywhere (the integration charms, the dev requirements, external user charms). I'll do the research now and switch over.

Reply:
Talked to ghislain, and decided to use dataclasses instead, which will fulfill the same goal with no deps.

Reply:
Ghislain and I talked, and decided that this wasn't worth the trouble.